RealmSwift/SwiftUI.swift

```diff
@@ -645,6 +645,285 @@ extension EnvironmentValues {
+///
+///         var body: some View {
+///            switch asyncOpen {
+///            case .notOpen:
+///                ProgressView()
+///            case .open(let realm):
+///                ListView()
+///                   .environment(\.realm, realm)
+///            case .error(_):
+///                ErrorView()
+///            case .progress(let progress):
+///                ProgressView(progress)
+///            }
+///         }
+///     }
+///
+/// This opened `realm` can be later injected to the view as an enviroment value which will be used by our property wrappers
+/// to populate the view with data from the opened realm
+///
+///     ListView()
+///        .environment(\.realm, realm)
+///
+@available(iOS 13.0, macOS 11.0, tvOS 13.0, watchOS 6.0, *)
+@propertyWrapper public struct AsyncOpen: DynamicProperty {
+    @Environment(\.realm) var realm
+    @ObservedObject private var storage = ObservableAsyncOpenStorage()
+
+    private func asyncOpenForUser<T: BSON>(_ user: User, partitionValue: T, configuration: Realm.Configuration) {
+        let userConfig = user.configuration(partitionValue: partitionValue, cancelAsyncOpenOnNonFatalErrors: true)
+        let userSyncConfig = userConfig.syncConfiguration
+        var configuration = configuration
+        configuration.syncConfiguration = userSyncConfig
+        storage.asyncOpen(configuration: configuration)
+            .sink { completion in
+                if case .failure(let error) = completion {
+                    self.storage.asyncOpenState = .error(error)
+                }
+                cancel()
+            } receiveValue: { realm in
+                self.storage.realm = realm
+                self.storage.asyncOpenState = .open(realm)
+            }.store(in: &storage.cancellables)
+    }
+
+    /// :nodoc:
+    public var projectedValue: Published<AsyncOpenState>.Publisher {
+        return storage.$asyncOpenState
+    }
+
+    /// :nodoc:
+    public var wrappedValue: AsyncOpenState {
+        storage.asyncOpenState
+    }
+
+    /**
+     This will cancel any notification from the property wrapper states
+     */
+    public func cancel() {
+        storage.cancellables.forEach { $0.cancel() }
+        storage.cancellables = []
+    }
+
+    /**
+     Initialise the property wrapper
+     - parameter appId: The unique identifier of your Realm app.
+     - parameter partitionValue: The `BSON` value the Realm is partitioned on.
+     - parameter baseConfiguration: A Base Configuration use when creating the Realm,
+       if empty the configuraion is setted to the `defaultConfiguration`
+     - parameter timeout: The maximum number of milliseconds to allow for a connection to
+       become fully established., if empty or `nil` no connection timeout is setted.
+     */
+    public init<T: BSON>(appId: String, partitionValue: T, baseConfiguration: Realm.Configuration = Realm.Configuration.defaultConfiguration, timeout: UInt? = nil) {
```

Comment:
We can probably derive AppId from the last used AppId as most applications will be singular. This would allow the property to potentially have a default value, though this may be a chunk more of work.

I would also change the name of baseConfiguration to configuration.

Reply:
mmmm maybe i'm not understanding this correctly but is there a way to retrieve the appId from the latest used one?

Reply:
It's currently not exposed in any way, but we internally have a cache of all Apps which have been initialized in the current run of the application. We could make the appId optional, and then if it's `nil` check if there's exactly one RLMApp in the cache, and if so use that (and otherwise throw an exception).

Reply:
ok I see that we internally cache all the apps in s_apps, this is private, should we expose this internally on `RLMApp_Private`?

Reply:
Exposing s_apps to RLMApp_Private.h seems reasonable enough.